File: 125.valid-palindrome.rb
#
# @lc app=leetcode id=125 lang=ruby
#
# [125] Valid Palindrome
#

# @lc code=start
# @param {String} s
# @return {Boolean}
def is_palindrome(s)
  s = s.downcase.gsub(/[^a-z0-9]/, '').downcase
  left = 0
  right = s.size - 1

  chars = s.chars
  while left < right
    if chars[left] != chars[right]
      return false
    end

    left += 1
    right -= 1
  end

  true
end
# @lc code=end

puts(is_palindrome('A man, a plan, a canal: Panama'))
puts(is_palindrome('race a car'))
puts(is_palindrome(' '))

File: 125.valid-palindrome.go
/*
 * @lc app=leetcode id=125 lang=golang
 *
 * [125] Valid Palindrome
 */
package main

import (
	"regexp"
	"strings"
)

// @lc code=start
func isPalindrome(s string) bool {
	re := regexp.MustCompile(`[^a-zA-Z0-9]+`)
	str := strings.ToLower(s)
	str = re.ReplaceAllString(str, "")

	chars := []rune(str)
	length := len(chars)

	for i := 0; i < length/2; i++ {
		left := chars[i]
		right := chars[length-i-1]
		if left != right {
			return false
		}
	}

	return true
}

// @lc code=end
